File: server/src/utils/oss.ts
import OSS from 'ali-oss';
import config from '../config';
import path from 'path';
import crypto from 'crypto';

let client: OSS | null = null;

function getClient(): OSS {
  if (!client) {
    const { region, accessKeyId, accessKeySecret, bucket } = config.oss;
    if (!accessKeyId || !accessKeySecret || !bucket) {
      throw new Error('OSS configuration is missing');
    }
    client = new OSS({
      region,
      accessKeyId,
      accessKeySecret,
      bucket,
      secure: true // Use HTTPS
    });
  }
  return client;
}

export async function uploadImageToOSS(buffer: Buffer, originalFilename: string): Promise<string> {
  try {
    const ossClient = getClient();
    
    // Generate a unique filename: ocr/{YYYY}/{MM}/{DD}/{random}_{timestamp}.ext
    const now = new Date();
    const year = now.getFullYear();
    const month = (now.getMonth() + 1).toString().padStart(2, '0');
    const day = now.getDate().toString().padStart(2, '0');
    
    const ext = path.extname(originalFilename) || '.jpg';
    const random = crypto.randomBytes(4).toString('hex');
    const timestamp = now.getTime();
    
    const objectName = `ocr/${year}/${month}/${day}/${timestamp}_${random}${ext}`;
    
    const result = await ossClient.put(objectName, buffer);
    
    // Return the path (object name) or full URL depending on your need.
    // Returning object name is safer if you want to sign URLs later, 
    // but for now let's return the object name to store in DB.
    return objectName;
  } catch (error) {
    console.error('OSS Upload Error:', error);
    throw error;
  }
}

File: server-go/utils/oss.go
package utils

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"path"
	"time"

	"bbflow-server/config"

	"github.com/aliyun/aliyun-oss-go-sdk/oss"
)

var ossClient *oss.Client
var ossBucket *oss.Bucket

func getOSSBucket() (*oss.Bucket, error) {
	if ossBucket != nil {
		return ossBucket, nil
	}

	cfg := config.AppConfig.OSS
	if cfg.AccessKeyID == "" || cfg.AccessKeySecret == "" || cfg.Bucket == "" {
		return nil, fmt.Errorf("OSS configuration is missing")
	}

	var err error
	endpoint := fmt.Sprintf("%s.aliyuncs.com", cfg.Region)
	ossClient, err = oss.New(endpoint, cfg.AccessKeyID, cfg.AccessKeySecret)
	if err != nil {
		return nil, err
	}

	ossBucket, err = ossClient.Bucket(cfg.Bucket)
	if err != nil {
		return nil, err
	}

	return ossBucket, nil
}

func UploadImageToOSS(data []byte, originalFilename string) (string, error) {
	bucket, err := getOSSBucket()
	if err != nil {
		return "", err
	}

	now := time.Now()
	year := now.Format("2006")
	month := now.Format("01")
	day := now.Format("02")

	ext := path.Ext(originalFilename)
	if ext == "" {
		ext = ".jpg"
	}

	randomBytes := make([]byte, 4)
	rand.Read(randomBytes)
	randomStr := hex.EncodeToString(randomBytes)

	objectName := fmt.Sprintf("ocr/%s/%s/%s/%d_%s%s", year, month, day, now.UnixMilli(), randomStr, ext)

	err = bucket.PutObject(objectName, bytes.NewReader(data))
	if err != nil {
		return "", err
	}

	return objectName, nil
}
